parquet-hadoop/src/main/java/org/apache/parquet/hadoop/RowGroupMerger.java

```diff
@@ -0,0 +1,634 @@
+
+  private final MessageType schema;
+  private final CodecFactory.BytesInputCompressor compressor;
+  private final ParquetProperties parquetProperties;
+
+  RowGroupMerger(MessageType schema, CompressionCodecName compression, boolean useV2ValueWriter) {
+    this(schema, new Configuration(), compression, useV2ValueWriter);
+  }
+
+  RowGroupMerger(MessageType schema, Configuration conf, CompressionCodecName compression, boolean useV2ValueWriter) {
+    this(schema, conf, compression, createParquetProperties(useV2ValueWriter));
+  }
+
+  RowGroupMerger(MessageType schema, Configuration conf, CompressionCodecName compression, ParquetProperties parquetProperties) {
+    this.schema = schema;
+    this.parquetProperties = parquetProperties;
+    this.compressor = new CodecFactory(conf, this.parquetProperties.getPageSizeThreshold()).getCompressor(compression);
+  }
+
+  /**
+   * Merges the row groups making sure that new row groups do not exceed the supplied maxRowGroupSize
+   *
+   * @param inputFiles      input files to merge
+   * @param maxRowGroupSize the max limit for new blocks
+   * @param writer          writer to write the new blocks to
+   * @throws IOException if an IO error occurs
+   */
+  void merge(List<InputFile> inputFiles, final long maxRowGroupSize, ParquetFileWriter writer) throws IOException {
+
+    SizeEstimator estimator = new SizeEstimator(compressor.getCodecName() != CompressionCodecName.UNCOMPRESSED);
+    MutableMergedBlock mergedBlock = null;
+    for (InputFile file : inputFiles) {
+      try (ParquetFileReader reader = ParquetFileReader.open(file)) {
+
+        for (BlockMetaData blockMeta : reader.getRowGroups()) {
+          PageReadStore group = reader.readNextRowGroup();
+          Preconditions.checkState(group != null,
+            "number of groups returned by FileReader does not match metadata");
+
+          if (mergedBlock != null && mergedBlock.getCompressedBytesSize() + estimator.estimate(blockMeta) > maxRowGroupSize) {
+            saveBlockTo(mergedBlock, writer);
+            mergedBlock = null;
+          }
+
+          if (mergedBlock == null && estimator.estimate(blockMeta) > maxRowGroupSize) {
+            //save it directly without re encoding it
+            saveBlockTo(ReadOnlyMergedBlock.of(blockMeta, group, schema, compressor), writer);
+            continue;
+          }
+
+          if (mergedBlock == null) {
+            mergedBlock = new MutableMergedBlock(schema);
+          }
+
+          long sizeBeforeMerge = mergedBlock.getCompressedBytesSize();
+          mergedBlock.merge(blockMeta, group);
+          //update our estimator
+          long currentBlockEffect = mergedBlock.getCompressedBytesSize() - sizeBeforeMerge;
+          estimator.update(currentBlockEffect, blockMeta);
+        }
+      }
+    }
+    if (mergedBlock != null)
+      saveBlockTo(mergedBlock, writer);
+  }
+
+
+  private static void saveBlockTo(MergedBlock block, ParquetFileWriter writer) {
```

Comment:
As bloom filter feature is implemented, I guess we need to take care of that too.

Reply:
Since we are using page level APIs of ParquetFileWriter(i.e _writeDataPage_ and _writeDataPageV2_), that is handled for us under the hood.